### PR TITLE
Fixed: Array values are wrapped with double quotes

### PR DESCRIPTION
**Pull request by MuratUrsavas**
## What does this PR do?
Fixes wrapping array values with extra double quotes.

## GitHub issue or Phabricator ticket number?
https://phorge.wireload.net/T7534

## Checklist before merging

- [x] If have added tests.
- [ ] Is this a new feature? If yes, please write one phrase about this update.
- [ ] I've attached relevant screenshots (if relevant).

### CODE REVIEW

File: Cargo.toml
Comment (sergey-borovkov) [inline Cargo.toml:3]: Why ?

File: src/cli.rs
Comment (sergey-borovkov) [inline src/cli.rs:1151]: why does it need to be converted here? And I don't see any reason why secret pairs need to be of this type. Especially in exposed API.

Any specific reason?

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:117]: why are we not using serde to deserialize string?

**Review by sergey-borovkov — CHANGES_REQUESTED** [review]

**Comment by sergey-borovkov** [comment]
Also there is no new test that tests for issue you fixed.

File: Cargo.toml
Comment (MuratUrsavas) [reply Cargo.toml:3]: Because the code has changed. We're using versions to tag a specific state of software. If that state changes, the version has to change.

Comment (sergey-borovkov) [reply Cargo.toml:3]: No. We bump version when we release new version, not for every fix (this is not even enough as the version is referenced in docker and somewhere else)

File: src/cli.rs
Comment (MuratUrsavas) [reply src/cli.rs:1151]: No specific reason actually. Was just quickly following the reference chain. Maybe overdone it :smile:

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: Well, that was the failing part. It cannot automatically create the new `MockData` from a string. I needed to create that middle step to fix it. If you have a more idiomatic idea, I'm all ears.

**Comment by MuratUrsavas** [comment]
> Also there is no new test that tests for issue you fixed.

Current tests already cover it. Maybe need to modify the current mock data a little bit.

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: What's the issue? It's not enought data for me that it's a failing part. But if some field fails to parse correctly you can always use custom deserializer - take a look at EdgeAppManifest for instance it does that a lot

**Comment by sergey-borovkov** [comment]
> > Also there is no new test that tests for issue you fixed.
> 
> Current tests already cover it. Maybe need to modify the current mock data a little bit.

We still should cover the exact failure we had. Those tests did not reveal the issue. I'd propose to have a test that checks that the whole output is correct by comparing with the whole output string typed out in the test

**Comment by MuratUrsavas** [comment]
> We still should cover the exact failure we had. Those tests did not reveal the issue. I'd propose to have a test that checks that the whole output is correct by comparing with the whole output string typed out in the test

Will do that with the mock data change.

File: src/cli.rs
Comment (sergey-borovkov) [reply src/cli.rs:1151]: I feel like it's a bit more natural for us not to expose those internal types they can be private to the module so that consumers don't have to care about it

File: Cargo.toml
Comment (MuratUrsavas) [reply Cargo.toml:3]: How about using dash notation for intermediary versions under development? Like `0.2.1-1` and while releasing we can get rid of the dash suffix.

File: src/cli.rs
Comment (MuratUrsavas) [reply src/cli.rs:1151]: Actually, the exposed part is still the same. Are you sure we're exposing the new type? Maybe I'm overseeing something?

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: I'd like to learn that custom deserializer. But we need to ship this one quickly. How about implementing it as a follow-up?

File: Cargo.toml
Comment (sergey-borovkov) [reply Cargo.toml:3]: I don't see any benefit to that as you can't download it from any place. And it's what we do in all our projects so what's the point of doing differently here?

File: src/cli.rs
Comment (sergey-borovkov) [reply src/cli.rs:1151]: EdgeAppServerValue::Str is exposed which is a private enum

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: ok since we need it fast - please create a GH issue and reference it there

File: Cargo.toml
Comment (sergey-borovkov) [inline Cargo.toml:3]: pls no :)

File: Cargo.toml
Comment (MuratUrsavas) [reply Cargo.toml]: We're testing it here and there and we should be providing it to other teams for tests before releasing it. And every software iteration has to carry a different tag. Reports have to include that version info to match the exact state of that software. You can use build number, commit hash, or anything else. But some info has to indicate that state. Version is the easiest one.

Comment (sergey-borovkov) [reply Cargo.toml]: When it's unreleased that's what you have git commit hash for. We are *not* providing that as a binary for testing to anyone.

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:288]: Let's test the whole file content, please. 

instead of contains using strict eq. 

I did this approach and it was an error since it turned out to cause bugs let's not use it further.

File: src/cli.rs
Comment (MuratUrsavas) [reply src/cli.rs]: Done!

File: Cargo.toml
Comment (MuratUrsavas) [reply Cargo.toml:3]: OK :)

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:290]: OK, let's do this one as a follow-up. I've done it but it kept invalidating the valid string generated by itself. Will take a deeper look after that. Approve and release as it is.

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:117]: Will do.

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:291]: don't check for negatives, please...

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:290]: I can't approve when the test for the actual content does not pass :( Maybe someone else will

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:208]: Clippy warning needs to be fixed

File: src/commands/edge_app_server.rs
Comment (renatgalimov) [inline src/commands/edge_app_server.rs:128]:
```suggestion
                        .map(|s| s.trim().replace(['"', '\''], ""))
```

File: src/commands/edge_app_server.rs
Comment (renatgalimov) [inline src/commands/edge_app_server.rs:207]:
```suggestion
                .map(|tag| format!(tag.to_string()))
```

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:117]: But there is no created issue and it's not referenced here :( So it's not resolved

File: src/commands/edge_app.rs
Comment (korvyashka) [inline src/commands/edge_app.rs:303]: Shouldnt fmt format this line?

File: src/commands/edge_app_server.rs
Comment (renatgalimov) [reply src/commands/edge_app_server.rs:290]: I agree with @sergey-borovkov, just do one equality test to a full file content.

File: src/commands/edge_app_server.rs
Comment (korvyashka) [inline src/commands/edge_app_server.rs:24]: Hm - inst this strictly cli related format, may be pass it with string to parse later?

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:207]: you don't need format! then though?

**Review by korvyashka — COMMENTED** [review]
Commented from mobile, so could miss or missunderstood smth.

1) please always do fmt and fix all clippy warnings
2) All tests on pr should pass before merging - not sure if they are, just saw the concersation

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:24]: yeah, I  also think it should all be in the server code.

File: src/commands/edge_app_server.rs
Comment (renatgalimov) [reply src/commands/edge_app_server.rs:207]: You are absolutely right

File: src/commands/edge_app.rs
Comment (MuratUrsavas) [reply src/commands/edge_app.rs:303]: Yeah forgot to run rustfmt while on my way out. Done.

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:116]: this function also needs to be tested especially considering that you parse it manually. You need to test it with invalid input/edge cases.

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: OK, how a test could fail even with its own output? Would you like me to wait to figure that bug out? I'm copying its left output to the right, and then assert it.

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:213]: Done!

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:122]: https://github.com/Screenly/cli/issues/119 :smile:

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done.

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:48]: Hmm. Maybe better to just use Vec<>? It preserves the order and we don't access the values by key/value

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:291]: maybe format it as multiline string to make it more readable?

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [inline src/commands/edge_app_server.rs:327]: Please improve the naming - see here how we name our tests https://phorge.wireload.net/w/developer_guidelines/tests/

should contain what it tests and the expected behavior (but see the document anyway for in depth explanation and examples)

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:48]: I think that could be left as it is.

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:291]: Done!

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: Done!

File: src/commands/edge_app_server.rs
Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:48]: We are not using it as a map but OK

**Review by sergey-borovkov — APPROVED** [review]

File: src/commands/edge_app_server.rs
Comment (StanIwan) [reply src/commands/edge_app_server.rs:48]: I'm with Sergey on this one, even though BTreeMap might work, if we can also just use Vec, that would be preferrable.

**Review by StanIwan — APPROVED** [review]
I approve, this are clippy warnings I get on 1.70:
```
siwan@siwan-x1-carbon:~/Projects/cli$ cargo clippy
warning: redundant clone
   --> src/commands/edge_app_server.rs:137:47
    |
137 |                     (k.clone(), Value::Array(v.clone()))
    |                                               ^^^^^^^^ help: remove this
    |
note: this value is dropped without further use
   --> src/commands/edge_app_server.rs:137:46
    |
137 |                     (k.clone(), Value::Array(v.clone()))
    |                                              ^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
    = note: `#[warn(clippy::redundant_clone)]` on by default

warning: called `.nth(0)` on a `std::iter::Iterator`, when `.next()` is equivalent
   --> src/commands/edge_app_server.rs:129:24
    |
129 |                     && v.chars().nth(0) == Some('[')
    |                        ^^^^^^^^^^^^^^^^ help: try calling `.next()` instead of `.nth(0)`: `v.chars().next()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_nth_zero
    = note: `#[warn(clippy::iter_nth_zero)]` on by default

warning: `cli` (bin "screenly") generated 2 warnings (run `cargo clippy --fix --bin "screenly"` to apply 2 suggestions)
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
```

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs:48]: Yeah, I know but there is a great possibility we could use it soon. Also, it doesn't create any dev overhead. That's why I opted to leave as it is.

Comment (sergey-borovkov) [reply src/commands/edge_app_server.rs:48]: @MuratUrsavas what possibility? I am not aware of it

File: Cargo.toml
Comment (renatgalimov) [reply Cargo.toml]: Guys. We use semantic versioning. We modify the version when we release. 
If you for whatever reason want to modify the version between releases you can use -rc tail.

It mains release candidate. 

2.0.0-rc1. 
2.0.0-rc2.
2.0.0-rc3.

We also can release the same version multiple times. We had to do it recently, to trigger the upgrade on UC16 devices. Then you can add -r tail

It means - revision.

2.0.0-r1 means 2.0.0 revision 1.

File: src/commands/edge_app_server.rs
Comment (korvyashka) [reply src/commands/edge_app_server.rs:48]: BTree is usually used in the databases(indexes) for fast reads(searches) specifically, there is usually no reason using it in the day to day code, unless there are a lot of items in the memory you want to search in fast. So I don't see any pros using it here.

**Review by renatgalimov — APPROVED** [review]

File: src/commands/edge_app_server.rs
Comment (MuratUrsavas) [reply src/commands/edge_app_server.rs]: OK, guys, you won. I've converted most of the BTreeMaps to Vec's but kept in YAML conversion. Without a custom deserializer, I cannot change it to a Vec. It will be done in another task.

**Comment by vpetersson** [comment]
Woop woop!

**Pull request merged**
